crates/codegen/syntax_templates/src/shared/text_index.rs

```diff
@@ -0,0 +1,67 @@
+use std::{
+    fmt::Display,
+    ops::{Add, AddAssign, Range},
+};
+
+use serde::Serialize;
+
+#[derive(Default, Copy, Clone, PartialEq, Eq, Debug, Serialize)]
+pub struct TextIndex {
+    pub utf8: usize,
+    pub utf16: usize,
+    pub char: usize,
+}
+
+impl PartialOrd for TextIndex {
```

Comment:
do we intend to expose similar utilities for typescript as well? If not, I wonder if we should make it private until we think about it.